packages/next/src/edge-function-source/get-edge-function.ts

```diff
@@ -108,29 +109,35 @@ export default function getNextjsEdgeFunction(
     }
 
     // Invoke the function injecting missing parameters
-    const result = await _ENTRIES[`middleware_${params.name}`].default.call(
-      {},
-      {
-        request: {
-          url: request.url,
-          method: request.method,
-          headers: toPlainHeaders(request.headers),
-          ip: header(request.headers, IncomingHeaders.Ip),
-          geo: {
-            city: header(request.headers, IncomingHeaders.City, true),
-            country: header(request.headers, IncomingHeaders.Country, true),
-            latitude: header(request.headers, IncomingHeaders.Latitude),
-            longitude: header(request.headers, IncomingHeaders.Longitude),
-            region: header(request.headers, IncomingHeaders.Region, true),
-          },
-          nextConfig: params.nextConfig,
-          page: pageMatch,
-          body: request.body,
-        },
-      }
+    const result = await withNextRequestContext(
+      { waitUntil: context.waitUntil },
+      () =>
+        _ENTRIES[`middleware_${params.name}`].default.call(
+          {},
+          {
+            request: {
+              url: request.url,
+              method: request.method,
+              headers: toPlainHeaders(request.headers),
+              ip: header(request.headers, IncomingHeaders.Ip),
+              geo: {
+                city: header(request.headers, IncomingHeaders.City, true),
+                country: header(request.headers, IncomingHeaders.Country, true),
+                latitude: header(request.headers, IncomingHeaders.Latitude),
+                longitude: header(request.headers, IncomingHeaders.Longitude),
+                region: header(request.headers, IncomingHeaders.Region, true),
+              },
+              nextConfig: params.nextConfig,
+              page: pageMatch,
+              body: request.body,
+            },
+          }
+        )
     );
 
-    context.waitUntil(result.waitUntil);
+    if (result.waitUntil) {
+      context.waitUntil(result.waitUntil);
+    }
```

Comment:
since we now have an "official" way of passing in waitUntil, i want to move away from the "returning a waitUntil promise on the result" inversion of control, because it leads to some contortions in practice when streaming is involved. so here i'm just making it so that we don't have to return a dummy promise anymore.